designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,442 @@
+
+::
+
+    $PKG_ROOT/build/
+        artifacts.json (not for MVP)
+        MyFunction1/
+             .... <build artifacts>
+
+          MyFunction2/
+             ... <build artifacts>
+
+          MyApiGw/
+            ... <build artifacts>
+
+          MyECRContainer/
+             ... <build artifacts>
+
+
+Future Extensions
+^^^^^^^^^^^^^^^^^
+In the future, we will change the limitation around each folder being named after the resource's LogicalID.
+Instead, we will support an artifacts.json file that will map Lambda function resource's LogicalId to the path to a
+folder that contains built artifacts for this function. This allows us to support custom build systems that use
+different folder layout.
+
+A well-known folder structure also helps “sam local” and “sam package” commands to automatically discover the
+built artifacts for each Lambda function and package it.
+
+Stable Builds
+~~~~~~~~~~~~~
+A build is defined to be stable if the built artifacts changes if and only if the contents or metadata
+(ex: timestamp, ownership) on source files changes. This is an important attribute of a build system. Since SAM CLI
+relies on 3rd party package managers like NPM to do the heavy lifting, we can only provide a "best effort" service here.
+By running ``sam build`` on a build system that creates a new environment from scratch
+(ex: Travis/CircleCI/CodeBuild/Jenkins etc), you can achieve truly stable builds. For more information on why this
+is important, refer to Debian's guide on `reproducible builds <https://reproducible-builds.org/>`_.
+
+
+SAM CLI does the following to produce stable builds:
+
+#. Clean build folder on every run
+#. Include metadata when coping files and folders
+#. Run build actions with minimal information passed from the environment
+
+
+Built-in Build Actions
+~~~~~~~~~~~~~~~~~~~~~~
+Build actions natively supported by SAM CLI follow a standard workflow:
+
+#. Search for a supported dependency manifest file. If a known manifest is not present, we will abort the build.
+#. Setup: Create build folder
+#. Resolve: Install dependencies
+#. Compile: Optionally, compile the code if necessary
+#. Copy Source: Optionally, Copy Lambda function code to the build folder
+
+Setup step is shared among all runtimes. Other steps in the workflow are implemented differently for each runtime.
+
+Javascript using NPM
+^^^^^^^^^^^^^^^^^^^^
+
+Install dependencies specified by ``package.json`` and copy source files
+
+**Manifest Name**: ``package.json``
+
++-------------+--------------------------------------+
+| Action      | Command                              |
++=============+======================================+
+| Resolve     | ``npm install``                      |
++-------------+--------------------------------------+
+| Compile     | No Op                                |
++-------------+--------------------------------------+
+| Copy Source | Copy files and exclude node_modules  |
```

Comment:
exclude node_modules? Not sure I understand the context here. Is this a guard to not copy node_modules if they installed them into their source directory? If so, why don't we have the same 'excludes' elsewhere?

Reply:
We should have similar excludes elsewhere. I will update the doc to reflect that